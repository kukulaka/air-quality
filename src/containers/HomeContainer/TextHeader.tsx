import React, { FC } from 'react';
import { Text, Box } from '@chakra-ui/react';


const TextHeader: FC= (Props) => {
  return (
    <>
      <Text textStyle="h1">Compare your Air</Text>
      <Text paddingTop="40px" id="header" paddingBottom="5px">
        Compare the air quality between cities in the UK.
      </Text>
      <Text>Select cities to compare using the search tool below.</Text>
    </>
  );
};

export default TextHeader;
