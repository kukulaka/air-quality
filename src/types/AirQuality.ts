export interface AirQuality {
  meta: MetaData;
  results: Location[];
  error: Error;
}

export interface Error {
  meta: null;
  results: null;
  error: boolean;
  errorMsg: string | null;
}

interface MetaData {
  name: string;
  license: string;
  website: string;
  page: number;
  limit: number;
  found: number;
}

interface LatLong {
  latitude: number;
  longitude: number;
}

interface Measurements {
  parameter: string;
  valu: number;
  lastUpdated: string;
  unit: string;
}
export interface Location {
  location: string;
  city: string;
  country: string;
  coordinates: LatLong;
  measurements: Measurements[];
}
