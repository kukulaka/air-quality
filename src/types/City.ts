export interface City {
  country: string;
  city: string;
  count: number;
  locations: number;
  firstUpdated: string;
  lastUpdated: string;
  parameters: string[];
  value: string;
  label: string;
}
