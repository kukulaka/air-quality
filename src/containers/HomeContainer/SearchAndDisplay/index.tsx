import React, { FC, useState } from 'react';
import AutoComplete from '../../../components/AutoComplete';
import { City } from '../../../types/City';
import ResultsLayout from '../../../components/ResultsLayout';

export interface Props {
  label: string;
  placeholder: string;
  itemsList: City[];
}

const SearchAndDisplay: FC<Props> = ({ label, placeholder, itemsList }) => {
  const [selectedItems, setSelectedItems] = useState<City[]>([]);

  const handleSelectedItemsChange = (selectedItems?: City[]) => {
    if (selectedItems) {
      setSelectedItems(selectedItems);
    }
  };

  return (
    <>
      <AutoComplete
        label={label}
        placeholder={placeholder}
        itemsList={itemsList}
        handleSelectedItemsChange={handleSelectedItemsChange}
        selectedItems={selectedItems}
      />
      <ResultsLayout selectedItems={selectedItems}/>
    </>
  );
};

export default SearchAndDisplay;
