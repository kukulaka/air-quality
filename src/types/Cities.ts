import { City } from './City';
export interface AllCities {
  meta: MetaData;
  results: City[];
  error: Error;
}

export interface CitiesList {
  cities: City[];
}

export interface Error {
  meta: null;
  results: null;
  error: boolean;
  errorMsg: string | null;
}

interface MetaData {
  name: string;
  license: string;
  website: string;
  page: number;
  limit: number;
  found: number;
}
