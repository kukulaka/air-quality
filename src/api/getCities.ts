import request from './request';
import { AllCities, Error } from '../types/Cities';

const getCities = async (): Promise<AllCities | Error | undefined> => {
  const cityURL: string =
    '/cities?limit=100&page=1&offset=0&sort=asc&country_id=GB&order_by=city';

  return await request('GET', cityURL, {})
    .then((response) => {
      return response.data as AllCities;
    })
    .catch((error) => {
      if (error) {
        return {
          error: true,
          errorMsg: error.message,
          meta: null,
          results: null,
        } as Error;
      }
    });
};

export default getCities;
