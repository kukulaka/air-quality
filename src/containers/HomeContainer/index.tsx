import React from 'react';
import PageContainer from '../../components/PageContainer';
import { Box, Spinner, Alert, AlertIcon } from '@chakra-ui/react';
import { AllCities, Error, CitiesList } from '../../../src/types/Cities';
import { City } from '../../../src/types/City';
import getCities from '../../api/getCities';
import TextHeader from './TextHeader';
import SearchAndDisplay from './SearchAndDisplay';


const HomeContainer: React.FC = (Props) => {
  const [allCitiesData, setAllCitiesData] = React.useState<Promise<
    AllCities | Error | undefined | null
  > | null>(null);
  const [error, setError] = React.useState<boolean>(false);
  const [loading, setLoading] = React.useState<boolean>(false);
  const [cityList, setCityList] = React.useState<CitiesList | null>(null);

  React.useEffect(() => {
    if (!allCitiesData) {
      setLoading(true);
      const allData = getCities();
      setAllCitiesData(allData);
      setLoading(false);
    }
  }, []);

  React.useEffect(() => {
    allCitiesData?.then((response) => {
      if (response!.results) {
        const cityArray: City[] = [];
        response!.results.map((city) => {
          const cityObj = {
            ...city,
            label: city.city,
            value: city.city,
          };
          cityArray.push(cityObj);
        });
        const cityUpdate: CitiesList = { cities: cityArray };
        setCityList(cityUpdate);
      }
      if (response!.error) {
        setError(true);
      }
    });
  }, [allCitiesData]);

  return (
    <PageContainer>
      <Box width="100%" height="100%" textAlign="center">
        <TextHeader />
        <Box
          paddingTop="30px"
          paddingBottom="30px"
          color="#000000"
          fontSize="16px"
          width="100%"
        >
          {loading ? (
            <Spinner />
          ) : (
            <>
              {error ? (
                <Alert status="error">
                  <AlertIcon />
                  There was an error retrieving cities data
                </Alert>
              ) : (
                <>
                  {cityList && cityList.cities.length > 0 && (
                    <SearchAndDisplay
                      label=""
                      placeholder="Type City name"
                      itemsList={cityList!.cities}
                    />
                  )}
                </>
              )}
            </>
          )}
        </Box>
      </Box>
    </PageContainer>
  );
};

export default HomeContainer;
