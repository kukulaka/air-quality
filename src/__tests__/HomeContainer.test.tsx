import * as React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import HomeContainer from '../containers/HomeContainer';
import { shallow, configure } from 'enzyme';
import getCities from '../api/getCities';
import { AllCities } from '../types/Cities';

// mock API call
configure({ adapter: new Adapter() });
jest.mock('../api/getCities');

describe('home container should retrieve city data on load', () => {
  it('t1', (done) => {
    const mockedData: AllCities = {
      meta: {
        name: 'test',
        license: 'test',
        website: 'test',
        page: 1,
        limit: 1,
        found: 1,
      },
      error: { meta: null, results: null, error: false, errorMsg: null },
      results: [
        {
          country: 'GB',
          city: 'Manchester',
          count: 1,
          locations: 1,
          firstUpdated: '2021-04-01T00:00',
          lastUpdated: '2021-04-01T00:00',
          parameters: ['p02', 'o2'],
          value: 'Manchester',
          label: 'Manchester',
        },
      ],
    };
    (getCities as jest.MockedFunction<typeof getCities>).mockResolvedValueOnce(
      mockedData
    );
    const wrapper = shallow(<HomeContainer></HomeContainer>);
    setImmediate(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });
});
