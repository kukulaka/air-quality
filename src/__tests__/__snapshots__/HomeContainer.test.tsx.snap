// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home container should retrieve city data on load t1 1`] = `ShallowWrapper {}`;
